src/Microsoft.SqlTools.ServiceLayer/LanguageExtensibility/ExternalLanguageOperations.cs

```diff
@@ -0,0 +1,348 @@
+                            });
+                        }
+                        ExternalLanguage metadata = dic[id];
+                        metadata.Contents.Add(new ExternalLanguageContent
+                        {
+                            EnvironmentVariables = envVariables,
+                            Parameters = parameters,
+                            Platform = platform,
+                            ExtensionFileName = extentionFileName
+                        });
+                    }
+                }
+            }
+            return new List<ExternalLanguage>(dic.Values);
+        }
+
+        private string GetCreateScript(ExternalLanguage language, Dictionary<string, object> parameters)
+        {
+            return GetLanguageModifyScript(language, language.Contents, parameters, ModifyType.Create);
+        }
+
+        private string GetUpdateScript(ExternalLanguage language, ExternalLanguageContent content, Dictionary<string, object> parameters, ContentModifyType contentModifyType)
+        {
+            return GetLanguageModifyScript(language, new List<ExternalLanguageContent> { content }, parameters, ModifyType.Alter, contentModifyType);
+        }
+
+        private string GetLanguageModifyScript(ExternalLanguage language, List<ExternalLanguageContent> contents, Dictionary<string, object> parameters, ModifyType modifyType, ContentModifyType contentModifyType = ContentModifyType.Add)
+        {
+            string contentScript = string.Empty;
+            for (int i = 0; i < contents.Count; i++)
+            {
+                var content = contents[i];
+                string seperator = contentScript == string.Empty ? "" : ",";
+                contentScript = $"{contentScript}{seperator}{GetLanguageContent(content, i, parameters)}";
+            }
+
+            string ownerScript = string.IsNullOrWhiteSpace(language.Owner) ? "" : $"AUTHORIZATION {language.Owner}";
+            string scriptAction = modifyType == ModifyType.Create ? CreateScript : AlterScript;
+            string contentAction = "FROM";
+            if (modifyType == ModifyType.Alter)
+            {
+                switch (contentModifyType)
+                {
+                    case ContentModifyType.Modify:
+                        contentAction = SetContentScript;
+                        break;
+                    case ContentModifyType.Add:
+                        contentAction = AddContentScript;
+                        break;
+                    case ContentModifyType.Remove:
+                        contentAction = RemoveContentScript;
+                        break;
+                }
+            }
+            return $@"
+{scriptAction} {EscapeString(language.Name)}
+{ownerScript}
+{contentAction} {contentScript}
+";
+        }
+
+        private string EscapeString(string value)
+        {
+            string escapedString = CUtils.EscapeStringSQuote(value);
```

Comment:
Why are you escaping both quotes and brackets? You should only be doing one based on what type of quoting is occurring - doing both is going to result in invalid scripts.

Reply:
yes I just realized I messed it up again, for when wrapping the name with "'" I have to use EscapeStringSQuote and whe wrap it with bracket, I have to use  EscapeStringCBracket

---
In reply to: [379498227](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r379498227) [](ancestors = 379498227)